arrow-fx-test/src/main/kotlin/arrow/test/laws/BracketLaws.kt

```diff
@@ -154,15 +168,62 @@ object BracketLaws {
         .equalUnderTheLaw(just(expected), EQ)
     }
 
-  fun <F> Bracket<F, Throwable>.guaranteeMustRunFinalizerOnError(
+  fun <F> Bracket<F, Throwable>.guaranteeCaseMustRunFinalizerOnError(
     EQ: Eq<Kind<F, Int>>
   ): Unit =
-    forAll(Gen.int(), Gen.throwable()) { i, t ->
+    forAll(Gen.int(), Gen.throwable().raiseError<F, Int, Throwable>(this)) { i, fe ->
       val msg = AtomicIntW(0)
-      raiseError<Int>(t)
-        .guarantee(unit().map { msg.value = i })
+      fe
+        .guaranteeCase { unit().map { msg.value = i } }
         .attempt()
         .map { msg.value }
         .equalUnderTheLaw(just(i), EQ)
     }
+
+  fun <F> Bracket<F, Throwable>.onErrorMustRunFinalizerOnError(
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(
+      Gen.throwable().raiseError<F, Int, Throwable>(this),
+      Gen.int()
+    ) { fe: Kind<F, Int>, i: Int ->
+      val msg = AtomicIntW(0)
+      fe
+        .onError { unit().map { msg.value = i } }
+        .attempt()
+        .map { msg.value }
+        .equalUnderTheLaw(just(i), EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.bracketShouldBeStackSafe(
+    iterations: Int,
+    EQ: Eq<Kind<F, Int>>
+  ) {
+
+    fun ioBracketLoop(i: Int): Kind<F, Int> =
+      unit().bracket(use = { just(i + 1) }, release = { unit() }).flatMap { ii ->
+        if (ii < iterations) ioBracketLoop(ii)
+        else just(ii)
+      }
+
+    forFew(100, Gen.int().applicativeError(this)) { fa ->
+      fa.flatMap { ioBracketLoop(0) }.equalUnderTheLaw(just(iterations), EQ)
+    }
+  }
+
+  fun <F> Bracket<F, Throwable>.guaranteeCaseShouldBeStackSafe(
+    iterations: Int,
+    EQ: Eq<Kind<F, Int>>
+  ) {
+    fun ioGuaranteeCase(i: Int): Kind<F, Int> =
+      unit().guaranteeCase { unit() }.flatMap {
+        val ii = i + 1
+        if (ii < iterations) ioGuaranteeCase(ii)
+        else just(ii)
+      }
+
+    forFew(100, Gen.int().applicativeError(this)) { fa ->
```

Comment:
Idem `forFew`

Reply:
Done in https://github.com/arrow-kt/arrow-fx/pull/44/commits/f90752b3455b61b9bcb439a2d18cbb0fd43d7f1c